### ISSUE TITLE
Operation Observers in 2.5

### ISSUE THREAD

**Issue by jshier** [danthorpe/Operations#152]
In 2.5 it was made a precondition failure to try to add an observer to an operation that had already progressed to the `.Ready` state. To me, this seems to greatly limit utility of observers, in some situations anyway. For instance, I can no longer decouple observation with operation creation. i.e. I can do this anymore:
```swift
var currentOperation: Operation? {
    didSet {
        guard let operation = currentOperation else { return }
        operation.addObserver(...)    
   }
}
...
func doSomething() -> Void {
    currentOperation = OperationManager.doSomething {
        print("did something")
    }
}
```
Instead, if I wish to add an observer, I must do it at operation creation, forcing me to tightly couple my observers and operations. This limits the scenarios in which I can use observers and the functionality of the observers to that which can be tightly coupled.

I'm not too familiar with the underly observer code, but perhaps it would be possible to, instead of checking to make sure the operation isn't yet ready before adding an observer, "play back" the state transitions the observer must logically have missed right after adding the observer to the operation. I realize this might not be possible for something like the finished or produced observers, depending on the states, but should be possible for general start observers and perhaps others.

**Comment by danthorpe** [comment]
Hi @jshier, hmm, but doesn't the use of `precondition` make explicit what was already the expectation of `Operation`?

If an observer is added after the operation moves past the `.Ready` state, then there is the potential that they will be added after the relevant event has passed. I will admit however, that given that now `OperationObserver` is somewhere redundant, that this only applies to some of the observer types, and that it could be improved, i.e. it should be possible to add an `OperationDidFinishObserverType` right up until it goes to `.Finishing` state.

Can you explain your usage a little bit more? What sort of observers are you adding, and how does the `OperationManager` work with regards to adding these operations to a queue versus adding the observer?

**Comment by jshier** [comment]
`OperationManager`s of the type I've been writing usually work like this:
```swift
class OperationManager : OperationQueue {
    private static let sharedManager = OperationManager()

    static func doSomething(completionHandler: () -> ()) -> SomeOperation {
        let operation = SomeOperation(completionHandler: completionHandler)
        sharedManager.addOperation(operation)

        return operation
    }
}
```
What I'd like to do, is base whatever behavior I can on the operation state, and observers are perfect for this. For instance, enabling or disabling UI, showing progress indicators, etc. Sometimes all in one observer, sometimes in observers for each element. (If you've ever used the old AFNetworking operations, the library also included categories for things like `UIRefreshControl` to be able to set its state using an operation. It did so by using key value observing.)

With 2.5, however, I have to pass the observer through to the manager, and can no longer use the property observers to set them up.
```swift
class OperationManager : OperationQueue {
    private static let sharedManager = OperationManager()

    static func doSomething(observer: StartFinishObserver, completionHandler: () -> ()) -> SomeOperation {
        let operation = SomeOperation(completionHandler: completionHandler)
        operation.addObserver(observer)
        sharedManager.addOperation(operation)

        return operation
    }
}
```
Which means I must create an API, wherever I use the observers, to create and pass them in:
```swift
func doSomething() {
    let observer = StartFinishObserver(...)
    currentOperation = OperationManager.doSomething(observer) {
        print("did something")
    } 
}
```
And this is really the best case scenario, where everything I have to do in the observer is right there already. In that case, I might as well just put the start code directly above the operation call and put the finish code in the completion handler, which is what I was trying to avoid in the first place. It also makes having multiple observers rather difficult, as the tight coupling would be multiplied.

**Comment by danthorpe** [comment]
Hi @jshier.

Okay, so before 2.5, you could call:

```swift
let operation = OperationManager.doSomething()
operation.addObserver(anObserver)
```
without any complaints?

There were two changes in 2.5. One was to switch to using `precondition` instead of `assert` and the other was to change the check to ` < .Ready` instead of ` < .Executing`.

The change to `.Ready` from `.Executing` was from @difujia, and his commit comment was:

> Improvements: conditions and observers should be add before operations is 'ready' instead of 'executing'.

> Despite the fact that the time window for state transition from 'ready' to 'executing' is rather short, 'ready' is the more accurate state for last chance of adding conditions and observers, because condition expresses 'readiness' and observer relies on 'readiness' to receive all life cycle events.

I definitely agree with this point regarding conditions, but can see that it is perhaps restrictive for observers.

I think it makes sense that `Operation` should provide guidance to consumers about when it is safe to add observers and conditions. 

But I guess it depends on the observer. At the moment, this is the code:

```swift
    public func addObserver(observer: OperationObserverType) {
        precondition(state < .Ready, "Cannot modify observers after operations has been ready, current state: \(state).")
        observers.append(observer)
    }
```

This treats all observers the same, despite them being triggered by different state change events. But, seeing as we can use method overloading... I could improve this to something like this:

```swift
    public func addObserver(observer: protocol<OperationDidStartObserver, OperationDidCancelObserver>) {
        precondition(state < .Ready, "Cannot modify observers after operations has been ready, current state: \(state).")
        observers.append(observer)
    }

    public func addObserver(observer: protocol<OperationDidProduceOperationObserver, OperationDidFinishObserver>) {
        precondition(state < .Executing, "Cannot modify observers after operations is executing, current state: \(state).")
        observers.append(observer)
    }

    public func addObserver(observer: OperationObserver) {
        precondition(state < .Executing, "Cannot modify observers after operations is executing, current state: \(state).")
        observers.append(observer)
    }  
```

The alternative is to remove the `precondition` checks, but add locks around the `observers` property - which would slow execution down.

Thoughts?

**Comment by difujia** [comment]
IMHO, `.Ready` is the only safe choice of the last chance allowing to add observers. The reason is simple, all operation life cycle events happen after `.Ready`, and their transitions are all unpredictable. Restricting the addition of observers before `.Ready` guarantees the observers being called if corresponding events occur. If the restriction did not exist, client code relying on execution of observers is likely to produce concurrency bugs.

**Comment by danthorpe** [comment]
I'm pretty much agree here with @difujia, but I'm a little concerned that something so trivial is the cause of any significant issues particularly if it has limited the flexibility of the framework. Partly I think the "flexibility" available prior to 2.5 was actually just dangerous/buggy code which we've now fixed/improved.

However, saying this, I think `OperationDidFinishObserver` is an edge case which could get some special treatment.

1. The finishing state transitions are very well defined, and it's impossible to prevent an operation from going to `.Finished` from `.Finishing`. It is also not possible to get to `.Finished` without going to `.Finishing` first. This is the transitions where these observers are triggered.

2. This is easily the most used and most useful observer. Given that we can now separate out `FinishedObserver` from the cover-all `BlockObserver` I think a special case overload for it would be okay which checks that the operation has not yet transitioned to `.Finishing`.

@jshier can you provide *even* more details about the architectural issue you are having here. In the programming guide, I've explicitly recommended as a "best practice" to set up dependencies, observers & conditions on operations before adding them to queues. My reasoning for this that you're far less likely to develop concurrency bugs and it will lead to code which is easier to read & maintain. I'm not sure whether to consider your `OperationManager` class as something many people will do (i.e. should be supported) or something which is perhaps an anti-pattern.

In a previous app I was working on I did have a similar "manager" style class for vending configured `Operation`s for various tasks. For example a `ContactManager`:

```swift

protocol ManagerType: class {

    static func start() -> NSOperation?

    var queue: OperationQueue { get }
    var readWriteConnection: YapDatabaseConnection { get }
}

class Manager: NSObject {

    static let queue = OperationQueue()

    let readWriteConnection = database.newConnection()

    var queue: OperationQueue {
        return Manager.queue
    }

    func enqueueOperation(operation: NSOperation, withCompletion completion: dispatch_block_t = { }) {
        operation.addCompletionBlock {
            dispatch_async(Queue.Main.queue, completion)
        }
        queue.addOperation(operation)
    }
}

class ContactManager: Manager, ManagerType {

    // etc

    // MARK: Promotion & Demotion
    func promotePerson(person: Person, automaticallyCreateReminder: Bool = true) -> PromotePersonOperation? {
        return PromotePersonOperation(person: person, automaticallyCreateReminder: automaticallyCreateReminder, readWriteConnection: readWriteConnection)
    }

    func demoteContact(contact: Contact) -> DemoteContractOperation? {
        return DemoteContractOperation(contact: contact, readWriteConnection: readWriteConnection)
    }
}
```

N.B. - I've left lots of details out of this, e.g. the `ContactManager` *is* follows a singleton pattern, but `Manager` does not.

In the app, generally this is used because it needs to manage some central resource (e.g. address book / contacts, cloud kit sync etc) but typically the functions which `doSomething` just return the configured operation. Enqueuing onto a queue would then be performed separately, using either the manager's queue with `enqueueOperation(: withCompletion:)` or another queue. This allows the opportunity to add more conditions / observers.

A key difference here, is that this is not a subclass of `OperationQueue`, which I have never considered doing before.

**Comment by jshier** [comment]
At a high level I guess my thinking is that the observers provided by `Operations` should be as flexible as KV observers you can use on regular `NSOperation`s. That is, observers can be added at any time, removed at any time, and don't care what state the operation is in when added. Observers provided by `Operations` should be superior because the developer doesn't need to know a key to observe, and specific, strongly typed closures can be used to run actions, rather than selectors. It also allows me to create custom observers which can return specific values from the operation in their closures, allowing me to pass state along without having to know anything about the underlying operation. 

More specifically, my ideal usage for observers is completely decoupled from the lifetime of the operation. That is, let's say I have a long running operation, like polling against a network endpoint that can take several minutes. In one view controller, I initiate the operation using a `Manager` class like I've already described. This class also tracks any long running operations, say by storing the operation in a property, `currentOperation`. Now say that, during this operation, various bits of the app can no longer be navigated to or perhaps some UI elements are disabled. In that case, as the user navigates around the app, the current view controller would ask the responsible manager for their active operation and add an observer to it so the controller's state would update automatically when the operation changes state. Additionally, the controller would be able to use the state captured by the observer, as described above, to update the UI with the new info when the operation completed.

Of course, much the same thing could be accomplished using `NSNotificationCenter`, where the controllers register and unregister for notifications broadcast by the operation. The biggest drawback in this architecture is getting new data at the end. Either the controller will have to make a separate request for the updated data it wants, or I'd have to find a way to encode and decode the data returned into an `NSNotification`, which isn't really something I want to put effort into. 

As for my use of `OperationQueue` subclasses as managers, I mainly do that to avoid having view controllers manage a queue and enqueue operations themselves. That just seems likely to pollute view controllers even more than they already are. Plus it's a place for operations to be visible to any controller that's interested, since my current app has a lot of long running operations, some of which affect various parts of the app.

**Comment by danthorpe** [comment]
Okay, these are all fair comments. I'm eager to see a resolution to this, so I'm going to provide overrides to `addObserver` which will sanity check based on what kind of observer protocol is being added.

@jshier I'm assuming from your usage case here, that you're adding observers for did finish?

**Comment by danthorpe** [comment]
Hi @jshier - I've made changes which (I think) strike a good balance between safety and flexibility regarding this issue. Merged into `development`. Let me know what you think so that we can resolve this issue. Cheers!

**Comment by jshier** [comment]
@danthorpe Sorry for the delay, I've started reviewing the observer architecture this weekend again. I think the changes so far are great. Since it's considered unsafe to add a start observer after the operations has actually started, what about something to tell an observer when it's been attached to an operation? In my use cases, it's not so much that an operation has started that important, it's that it hasn't yet finished. I'm hesitant to suggest yet another observer protocol, so I'm not sure where it would live, but I think this sort of functionality would be the final piece of the puzzle for me here. If you need more detail I can sketch something out, but I'm interested if you think this would be valuable or how you'd think it would work best.

**Comment by danthorpe** [comment]
@jshier Interesting. That's a possibility I hadn't considered...

At the moment there are four observer protocols which all extend from `OperationObserverType` which is an empty protocol - this could have a function added:

```swift
public protocol OperationObserverType {
    func didAttachToOperation(operation: Operation)
}
```

However, for your use case:

>  it's not so much that an operation has started that important, it's that it hasn't yet finished

Does this mean that you can use a type which conforms to `OperationDidFinishObserver`? If so, then it can be added right up until the operation gets to `.Finishing` state (i.e. even when the operation is `.Executing`. Will this not fulfill your requirements?

**Comment by jshier** [comment]
What I meant was that, for my purposes, part of my code doesn't care so much whether the operation has specifically started, just that it exists and hasn't finished yet. I suppose just using a finish observer could do what I want, but it would mean some of my code would have to live outside the observer. For example:
```swift
// In some view controller.
var operation: SomeOperation? {
    didSet {
        // Setup UI for active operation
        let observer = FinishedObserver { (operation, error) in
            // Setup UI for finished operation
        }
        operation.addObserver(observer)
    }
}
```
vs.
```swift
// In some view controller.
var operation: SomeOperation? {
    didSet {
        let observer = AttachAndFinishedObserver(attachBlock: { operation in
            // Setup UI for active operation
        }) { (operation, error) in
            // Setup UI for finished operation
        }
        operation.addObserver(observer)
    }
}
```
To me, it's somewhat cleaner to have the before and after code both live in the observer, rather than some of it outside and some of it inside. Ultimately though, I guess it doesn't make that much difference. 

After thinking about this scenario more, my main concern is that, unless I'm still adding all of my observers up front, there's no way for me to safely add observers. For example, if I'm attaching to operations on the fly, there's no way to be sure I won't ever hit one of the state assertions, since I have no access to the operation's `state` in order to check it's safe to attach. So the changes made haven't really changed the behavior I was concerned about. I will eventually see a crash if I add finish observers to running operations. It's just inevitable over time, and over a large enough number of users, there will be crashes.

So I guess we're back to square one. There seem to be several, sometimes competing, requirements for a observer system.

1. Simplicity. There are likely only a few interesting events in an `Operation`'s lifetime, so any observation API should be limited to those.
2. Safety It should be safe to add observers at any time. If it's not, the API should enforce a contract around safe behavior.
3. Performance. Adding and running observers should have as little impact on the CPU or memory performance of the `Operation`. 
4. Flexibility. A useful observation system allows users to add or remove observers at any time.
5. Customizability. The base observers offered by `Operation` should be supported by an infrastructure or common code that can be used by custom `Operation`s to offer new observation points.

Of course, I just made these requirements up, but I think they provide a good overview. The current architecture meets 1, 2, and 3 pretty well, but offers little around 4 and 5, which is where observers could really enable interesting features.

[Mike Ash wrote an interesting article a while back about Swift observers](https://www.mikeash.com/pyblog/friday-qa-2015-01-23-lets-build-swift-notifications.html). It's an interesting take on the issue. 

In conclusion, I think a more fundamental change to the observer architecture will be required to fulfill my dream of being able to attach observers at any point.

**Comment by danthorpe** [comment]
Hi @jshier, as always, these are really good points and very constructive.

I'll read what Mike has to say (I think I actually missed this post - so thanks), and consider some more fundamental architectural changes. My hunch at the moment is that some thread safety can deliver 4 & 5 from your list without impacting 3 too much.

**Comment by danthorpe** [comment]
@jshier - can we re-visit this issue considering the recent post-v2.6 changes to observers? Did we address all your concerns? If so, can we close this?

**Comment by jshier** [comment]
I'll take another look tonight at the latest changes and see.

**Comment by danthorpe** [comment]
Thanks! (I'm just trying to manage the issue count)

**Comment by jshier** [comment]
Looks like this latest set of changes has made conforming to multiple observation protocols somewhat more complicated. I has this observer before:
```swift
struct NetworkNotifier: OperationStartNotifier, OperationFinishNotifier {
    static let didStartKey = "NetworkOperationDidStart"
    static let didFinishKey = "NetworkOperationDidFinish"
    
    let startKey = NetworkNotifier.didStartKey
    let finishKey = NetworkNotifier.didFinishKey
}

protocol OperationStartNotifier: OperationDidStartObserver {
    var startKey: String { get }
}

extension OperationStartNotifier {
    func didStartOperation(operation: Operation) {
        NotificationCenter.postNotification(startKey)
    }
}

protocol OperationFinishNotifier: OperationDidFinishObserver {
    var finishKey: String { get }
}

extension OperationFinishNotifier {
    func didFinishOperation(operation: Operation, errors: [ErrorType]) {
        NotificationCenter.postNotification(finishKey)
    }
}
```
Which now doesn't work because the compiler now sees two `events` properties, from the underlying protocol extensions. Adding my own `events` property doesn't work because the various `OperationObserverEvents` aren't public. Once they're made public, however, the observer works fine, once I add the `events` property. However, since it's not used at the moment, it may be better hold off on that feature/requirement, as right now it's pain for no gain. But it's up to you. I like the addition of the labels and `didAttach` conformance, as well as the enhancements to `BlockOperation`. 

The `didAttach` implementation is a little awkward, but does let me use a `BlockOperation` instead of writing my own:
```swift
extension UIActivityIndicatorView {
    func observe(operation: Operation?) {
        var observer = BlockObserver { [weak self] (operation, errors) in
            dispatchMain { self?.stopAnimating() }
        }
        observer.didAttachToOperation = { [weak self] operation in
           dispatchMain { self?.startAnimating() }
        }
        
        operation?.addObserver(observer)
    }
}
```

Overall, looks good. I'll open a new issue if I run into issues with the new implementation, though it'll likely be after 2.7 ships, as most of my effort is going into a project that ships in March, so I may not get around to testing `development` that often.

**Comment by jshier** [comment]
Somewhat tangential, but perhaps related: is there a way to check if, in that `UIActivityIndicatorView` extension above, any observer is already attached? I don't think there is, at least with a `BlockOperation`. Perhaps a custom observer type that stores a weak reference to the view, so I could check `operation.observers` if it already exists?

**Comment by danthorpe** [comment]
Okay. Will remove the events stuff.

**Comment by danthorpe** [comment]
Yeah, there is no way to check, and I can't introduce Equatable as that will force generics and screw everything up.

I think you'll need a custom protocol, which is Equatable then flatMap on the observers array, then find using a key of some sort.

**Comment by danthorpe** [comment]
Okay - events stuff has been removed. Gonna close this for now. Have created #202 for new observer type: `UIActivityIndicatorViewObserver`.
